pallets/fusion/src/lib.rs

```diff
@@ -0,0 +1,1989 @@
+		/// User has too much unbonding chunks
+		MaxUnbondingChunksExceeded,
+		/// No funds are available to withdraw
+		NoFundsToWithdraw,
+		/// Caller is not authorized for this operation
+		NotAuthorized,
+		/// No rewards were found for the era
+		NoRewardsForEra,
+		/// The exposure is not founds
+		ExposureNotFound,
+		/// The user was not found in the exposure
+		UserNotFoundInExposure,
+		/// Rewards were already claimed for this user / era
+		AlreadyClaimed,
+		/// A user tried to unbond another user but it's only allowed if the pool is destroying
+		PoolIsNotDestroying,
+		/// Action is not allowed as the pool is destroying
+		PoolIsDestroying,
+		/// To handle compounding and easy bouding and unbonding, avail currency has no minimum
+		NoMinAmountForAvailCurrency,
+		/// There is no controller address to withdraw to
+		NoControllerAddressForUser,
+		/// If you're active points are below minimum, you cannot set compound to true
+		CannotSetCompoudingWithLessThanMinimum,
+		/// The state cannot be set to open if the pool is not nominating
+		PoolIsNotNominating,
+		/// We cannot set the pool to have no nomination if it's active
+		CannotSetNoNominationsWhenPoolActive,
+		/// The fusion currency rate was not found
+		FusionCurrencyRateNotFound,
+		/// The fusion pallet is paused so the operation is not allowed
+		FusionPalletPaused,
+		/// Temp
+		CannotDepositAvailCurrency,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// TODO - Dummy extrinsic to add currency without bridge, to be removed
+		#[pallet::call_index(99)]
+		#[pallet::weight(T::WeightInfo::create_fusion_currency())]
+		pub fn deposit_currency_dummy(
+			origin: OriginFor<T>,
+			evm_address: EvmAddress,
+			currency_id: CurrencyId,
+			amount: FusionCurrencyBalance,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			Self::do_deposit_currency(evm_address, currency_id, amount)?;
+			Ok(())
+		}
+
+		/// Pauses the entire pallet, halts rewards creation. Claiming, unbonding and withdrawing is still possible
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::create_fusion_currency())]
+		pub fn pause(origin: OriginFor<T>) -> DispatchResult {
+			ensure_root(origin)?;
+
+			Self::do_pause();
+
+			Ok(())
+		}
+
+		/// Fills the funds account with the specified amount of funds.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::create_fusion_currency())]
+		pub fn fill_funds_account(origin: OriginFor<T>, amount: BalanceOf<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			T::Currency::transfer(
+				&who,
+				&Self::accounts().funds_reward_account,
+				amount,
+				ExistenceRequirement::KeepAlive,
+			)?;
+
+			Self::deposit_event(Event::FundsAccountFilled {
+				sender: who,
+				amount,
+			});
+
+			Ok(())
+		}
+
+		/// Creates a new currency
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::create_fusion_currency())]
+		pub fn create_fusion_currency(
+			origin: OriginFor<T>,
+			currency_id: CurrencyId,
```

Comment:
Is there a reason why the caller should be able to define the currency_id? Could be simpler to just fetch it from `FusionCurrencyCount`

Reply:
No reason in particular except simplicity and chosing relevant id compared to existing stuff on EVM
You think it should be changed ?

Reply:
If there is a need for it to be user defined then it's fine, otherwise I don't see why we would bother the user to define it

Reply:
Since it's MVP, let's leave it like this then maybe change it after demo'ing it

Reply:
Also, it allows us to create a currency with an id and in the same batch nominating from it and setting it top open instead of waiting